### PR TITLE
Fix typing issue in `MetaData.reflect()` in AsyncIO context

### PR DESCRIPTION
**Pull request by francipvb**
I've added some overloads to the method.

### Description
Mypy threats unions and overloads different. To fix the issue, I've added two overloads to the method.

### Checklist

This pull request is:

- [x] A documentation / typographical / small typing error fix
	- Good to go, no issue or tests are needed
- [x] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

**Comment by francipvb** [comment]
CC @CaselIT @zzzeek

**Comment by francipvb** [comment]
I added some additional mypy tests for `create_all()` and `drop_all()` but there are no issues for them.

**Comment by CaselIT** [comment]
Honestly this seems like a mypy bug... 

Looking at their tracker does not seem to be turning up anything. I'll try reporting it

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision d7be4309e2102e829b42bbd8fda7185f09392d73 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change d7be4309e2102e829b42bbd8fda7185f09392d73: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5242

**Comment by CaselIT** [comment]
I don't understand why drop_all and create_all are ok, but reflect is not

File: test/typing/plain_files/ext/asyncio/engines.py [hunk 1]
```diff
@@ -71,3 +72,10 @@ async def asyncio() -> None:
     ce.statement
     cc = select(1).compile(conn)
     cc.statement
+
+    async with e.connect() as conn:
+        metadata = MetaData()
+
+        await conn.run_sync(metadata.create_all)
+        await conn.run_sync(metadata.reflect)
+        await conn.run_sync(metadata.drop_all)
```
Comment (sqla-tester) [inline test/typing/plain_files/ext/asyncio/engines.py:81]: **Michael Bayer** (zzzeek) wrote:

can we also add table.create, table.drop, index.create, index.drop, and anything else where we have an engine argument.  should be a bunch of things in schema.py

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5242

**Review by sqla-tester — COMMENTED** [review]
**Michael Bayer** (zzzeek) wrote:

id like to add more test cases

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5242

**Comment by CaselIT** [comment]
> I don't understand why drop_all and create_all are ok, but reflect is not

found the reproducer, the issue is that mypy does not like the `**kw: Any`. Will report an issue.

See example:
```py
from typing import Any, Callable, Concatenate, ParamSpec, TypeVar
P = ParamSpec("P")
T = TypeVar("T")

def go(x: Callable[Concatenate[str, P], T], *a: P.args, **k: P.kwargs) -> T:
    return x("foo", *a, **k)

def ok(a: str | int, one: int) -> int:
    return 1

def error(a: str | int, one: int, **kw: Any) -> int:
    return 2

go(ok, 1)
go(error, 1)
```

**Comment by CaselIT** [comment]
that's wired, the above reproduces in one env but not in another. Same mypy version..

**Comment by CaselIT** [comment]
ok mypy errors only when strict is used

**Comment by CaselIT** [comment]
reported here: https://github.com/python/mypy/issues/17093

I think we should leave a comment to the overload referencing that issue

File: test/typing/plain_files/ext/asyncio/engines.py (hunk 1 above)
Comment (francipvb) [reply test/typing/plain_files/ext/asyncio/engines.py:81]: Sounds good. I'll add these.

**Comment by francipvb** [comment]
I've added tests for table, index and enum.

Because these are just typing related tests, I've re-used metadata object attributes to perform the tests, except for the enum type.

As constraints are attached to the table itself, I didn't found a way to test creation an dropping of these things.

Any suggestion? Should I leave this?

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 49e10e0d2a7bbadf471212033e25b7616b99c092 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 49e10e0d2a7bbadf471212033e25b7616b99c092 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5242

File: test/typing/plain_files/ext/asyncio/engines.py [hunk 2]
```diff
@@ -71,3 +73,25 @@ async def asyncio() -> None:
     ce.statement
     cc = select(1).compile(conn)
     cc.statement
+
+    async with e.connect() as conn:
+        metadata = MetaData()
+
+        await conn.run_sync(metadata.create_all)
+        await conn.run_sync(metadata.reflect)
```
Comment (sqla-tester) [inline test/typing/plain_files/ext/asyncio/engines.py:81]: **Michael Bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5242

**Review by sqla-tester — COMMENTED** [review]
**Michael Bayer** (zzzeek) wrote:

it's gerat.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5242

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5242 has been **merged**. Congratulations! :)

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5263 has been **merged**. Congratulations! :)

**Pull request closed without merging**